ui/library/src/context/ScrollContext.ts

```diff
@@ -39,8 +64,121 @@ export function useScrollContextProps(): IScrollContext {
     };
   }, [scrollRoot]);
 
+  // Causes the IntersectionObservers to disconnect and be recreated (useful when DOM changes)
+  const [observerIndex, setObserverIndex] = React.useState(0);
+  const resetScrollObservers = React.useCallback(() => {
+    setObserverIndex(observerIndex + 1);
+  }, [observerIndex]);
+
+  const [visibleOutlineTargets, setVisibleOutlineNodes] = React.useState<Set<NodeDestination>>(
+    () => {
+      const set = new Set<NodeDestination>();
+      Object.freeze(set);
+      return set;
+    }
+  );
+
+  const [visiblePageNumbers, setVisiblePageNumbers] = React.useState<Set<number>>(() => {
+    const set = new Set<number>();
+    Object.freeze(set);
+    return set;
+  });
+
+  const isOutlineTargetVisible = React.useCallback(
+    (dest: NodeDestination): boolean => {
+      return visibleOutlineTargets.has(dest);
+    },
+    [visibleOutlineTargets]
+  );
+
+  const isPageVisible = React.useCallback(
+    ({ pageNumber, pageIndex }: { pageNumber?: number; pageIndex?: number }): boolean => {
+      if (typeof pageIndex === 'number') {
+        pageNumber = pageIndex + 1;
+      }
+      if (typeof pageNumber !== 'number') {
+        return false;
+      }
+      return visiblePageNumbers.has(pageNumber);
+    },
+    [visiblePageNumbers]
+  );
+
+  const scrollToPage = React.useCallback(
+    ({ pageNumber, pageIndex }: { pageNumber?: number; pageIndex?: number }): void => {
+      if (typeof pageNumber === 'number') {
+        pageIndex = pageNumber - 1;
+      }
+      if (typeof pageIndex !== 'number') {
+        return;
+      }
+      document
+        .getElementById(generatePageIdFromIndex(pageIndex))
+        ?.scrollIntoView({ behavior: 'smooth' });
+    },
+    []
+  );
+
+  // Watch outline nodes
+  React.useEffect(() => {
+    const root = scrollRoot || document.documentElement;
+    const detector = new VisibilityDetector<NodeDestination>({
+      root: root,
+      setEntries: setVisibleOutlineNodes,
+      onVisibilityChange: ({ visibleElements, hiddenElements, lastEntries }) => {
+        const newEntries = new Set(lastEntries);
+        for (const el of hiddenElements) {
+          const dest = el.getAttribute('data-outline-target-dest');
```

Comment:
After changing the type of `visibleElements` and `hiddenElements` from `Element[]` to `IntersectionObserverEntry[]`, you will need to change this line, line 135, line 156 and line 161 as below:
```
const dest = el.target.getAttribute('data-outline-target-dest');
```
There you can get the correct targeted element that has changed